pkg/oidc/provider/authorize.go

```diff
@@ -70,14 +72,65 @@ func (h *authorizeHandler) authEndpoint(w http.ResponseWriter, r *http.Request)
 		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("error parsing parameters from request %v", err), http.StatusBadRequest, w)
 		return
 	}
+	token, err := h.getAndVerifyRancherTokenFromRequest(r)
+	// redirect to the login page if the token is not present or there is any error fetching it. We need to pass all the oidc parameters from the original request.
+	if err != nil {
+		u, err := url.Parse(settings.ServerURL.Get() + "/dashboard/auth/login")
+		if err != nil {
+			oidcerror.WriteError(oidcerror.InvalidRequest, "error parsing server url", http.StatusInternalServerError, w)
+			return
+		}
+		q := url.Values{}
+		q.Set("response_type", params.responseType)
+		q.Set("client_id", params.clientID)
+		q.Set("redirect_uri", params.redirectURI)
+		q.Set("scope", strings.Join(params.scopes, " "))
+		q.Set("code_challenge", params.codeChallenge)
+		if params.state != "" {
+			q.Set("state", params.state)
+		}
+		if params.nonce != "" {
+			q.Set("nonce", params.nonce)
+		}
+		u.RawQuery = q.Encode()
+
+		http.Redirect(w, r, u.String(), http.StatusFound)
+		return
+	}
+
 	// validate all parameter as per the oidc spec.
 	if params.redirectURI == "" {
 		oidcerror.WriteError(oidcerror.InvalidRequest, "missing redirect_uri", http.StatusBadRequest, w)
 		return
 	}
-	if _, err := url.Parse(params.redirectURI); err != nil {
+	redirectURL, err := url.Parse(params.redirectURI)
+	if err != nil {
 		oidcerror.WriteError(oidcerror.InvalidRequest, "invalid redirect_uri", http.StatusBadRequest, w)
 	}
+	if redirectURL.Host == r.URL.Host && redirectURL.Path == r.URL.Path {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "redirect_uri can't be the same as the host uri", http.StatusBadRequest, w)
+	}
+	oidcClients, err := h.oidcClientCache.GetByIndex(oidcClientByIDIndex, params.clientID)
+	if err != nil {
+		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("error retrieving OIDC client: %v", err), http.StatusBadRequest, w)
+		return
+	}
+	if len(oidcClients) == 0 {
+		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("OIDC client not found: %v", err), http.StatusBadRequest, w)
+		return
+	}
+	if len(oidcClients) > 1 {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "multiple OIDC clients with the same clientID found", http.StatusBadRequest, w)
+		return
+	}
+	oidcClient := oidcClients[0]
+	if !slices.Contains(oidcClient.Spec.RedirectURIs, params.redirectURI) {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "redirect_uri is not registered", http.StatusBadRequest, w)
+		return
+	}
+	if r.URL.Host == redirectURL.Host {
+		w.Header().Set("Access-Control-Allow-Origin", redirectURL.Host)
+	}
```

Comment:
The `if` may be redundant based on the check already in place for `redirectURI`.

```suggestion
	w.Header().Set("Access-Control-Allow-Origin", redirectURL.Host)
```

Reply:
On a second thought, would we still need this after `addHeadersMiddleware` was introduced?

Reply:
we check all redirectURLs from all `OIDCClients` in the `addHeadersMiddleware`. Here we override with the redirectURL for this specific `OIDCClient` (that's not possible in other endpoints as we don't know the exact `OIDCClient`)

Reply:
you're right we don't need the if here